TWTValidation/JSON Validator/Parser/TWTJSONSchemaParser.m

```diff
@@ -0,0 +1,652 @@
+    return nil;
+}
+
+
+/* Keyword: dependencies
+ Showing example of 1) property dependency 2) schema dependency; note these are equivalent for validation
+ 1) Property dependency, i.e., "dependencies" object's value is an array
+ { "type": "object",
+ "properties": {
+ "name": { "type": "string" },
+ "credit_card": { "type": "number" },
+ "billing_address": { "type": "string" } },
+ "required": ["name"],
+ "dependencies": {
+ "credit_card": ["billing_address"] }
+ }
+
+ 2) Schema dependency, i.e., "dependencies" object's value is a schema
+ { "type": "object",
+ "properties": {
+ "name": { "type": "string" },
+ "credit_card": { "type": "number" } },
+ "required": ["name"],
+ "dependencies": {
+ "credit_card": {
+ "properties": {
+ "billing_address": { "type": "string" }
+ },
+ "required": ["billing_address"] }
+ }
+ }
+ */
+- (NSArray *)parseDependenciesWithSchema:(NSDictionary *)schema
+{
+    NSDictionary *dependencies = schema[kTWTJSONSchemaKeywordDependencies];
+    if (!dependencies) {
+        return nil;
+    }
+    [self pushPathComponent:kTWTJSONSchemaKeywordDependencies];
+    [self checkObject:dependencies isOfClass:[NSDictionary class] allowNil:NO];
+
+    NSMutableArray *dependencyNodes = [[NSMutableArray alloc] init];
+    [dependencies enumerateKeysAndObjectsUsingBlock:^(NSString *key, id object, BOOL *stop) {
+        [self pushPathComponent:key];
+
+        [self checkObject:key isOfClass:[NSString class] allowNil:NO];
+
+        if ([object isKindOfClass:[NSDictionary class]]) {
+            TWTJSONSchemaDepedencyASTNode *node = [[TWTJSONSchemaDepedencyASTNode alloc] initWithKey:key valueSchema:[self parseSchema:object]];
+            [dependencyNodes addObject:node];
+        } else {
+            NSArray *propertySet = [self parseArrayOfUnqiueStringsFromSchema:dependencies forKey:key];
+            TWTJSONSchemaDepedencyASTNode *node = [[TWTJSONSchemaDepedencyASTNode alloc] init];
+            node.key = key;
+            node.propertySet = propertySet;
+            [dependencyNodes addObject:node];
+        }
+
+        [self popPathComponent];
+
+    }];
+
+    [self popPathComponent];
+    return dependencyNodes;
+}
+
+
+// Keyword: items, not
+// Example: { "items" : {} }
+- (TWTJSONSchemaASTNode *)parseSchemaFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    NSDictionary *nestedSchema = schema[key];
+    if (!nestedSchema) {
+        return nil;
+    }
+
+    [self pushPathComponent:key];
+    TWTJSONSchemaASTNode *node = [self parseSchema:nestedSchema];
+    [self popPathComponent];
+    return node;
+}
+
+
+#pragma mark - Helper methods
+
+- (void)checkObject:(id)object isOfClass:(Class)className allowNil:(BOOL)allowNil
+{
+
+    if (allowNil && !object) {
+        return;
+    }
+
+    if ([object isKindOfClass:className]) {
+        return;
+    }
+
+    NSString *reason = [NSString stringWithFormat:@"Invalid type encountered. Expected object of class %@.", NSStringFromClass(className)];
```

Comment:
You actually don’t need to call `NSStringFromClass`. Classes are objects that respond to `description`, so you can use the class like any other object.